modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -188,6 +216,9 @@
     /** Disconnected flag. */
     private volatile boolean disconnected;
 
+    /** Keeps histograms to measure durations of service methods. Guided by service name, then method name. */
+    private final Map<String, Map<String, MethodHistogramHolder>> invocationHistograms = new HashMap<>(1);
```

Comment:
@ivandasch, I considered using `ConcurrentHashMap`. I didn’t see a case for it. Getting histogram should be quick. You can find there are no iteration over the map and parallel puts/removes. The map isn’t available outside. I think it can occur that service method might be concurrently invoked when service is being undeployed. It would lead we get null-histogram. There is npe-protection for this case in `GridServiceProxy.callServiceMethod()`: 
```
                if (histogram != null)
                    histogram.value(duration);
```

Reply:
Event if getting histogram is quick, this fact doesn't make incorrect code correct. There is no guarantee, that after publication in one thread (put), you will see any result (get) in another. Even after a some long period of time.